target/kubernetes.go

```diff
@@ -0,0 +1,173 @@
+package target
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	dto "github.com/prometheus/client_model/go"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type KubernetesEndpointFetcher struct {
+	endpointname string
+	namespace    string
+
+	port   int
+	path   string
+	scheme string
+
+	client    *http.Client
+	authToken string
+
+	kube client.Client
+
+	clock           func() time.Time
+	refreshInterval time.Duration
+	mutex           sync.Mutex
+	cache           map[string][]dto.MetricFamily
+	lastUpdated     time.Time
+}
+
+type KubernetesEndpointFetcherOpts struct {
+	Endpointname string
+	Namespace    string
+
+	Port   int
+	Path   string
+	Scheme string
+
+	AuthToken          string
+	RefreshInterval    time.Duration
+	InsecureSkipVerify bool
+}
+
+func NewKubernetesEndpointFetcher(opts KubernetesEndpointFetcherOpts) (*KubernetesEndpointFetcher, error) {
+	restConf, err := ctrl.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	kubeClient, err := client.New(restConf, client.Options{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &KubernetesEndpointFetcher{
+		endpointname: opts.Endpointname,
+		namespace:    opts.Namespace,
+		port:         opts.Port,
+		path:         opts.Path,
+		scheme:       opts.Scheme,
+
+		client: &http.Client{
+			Timeout: 5 * time.Second,
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: opts.InsecureSkipVerify,
+				},
+			},
+		},
+		authToken: opts.AuthToken,
+
+		kube: kubeClient,
+
+		refreshInterval: opts.RefreshInterval,
+		mutex:           sync.Mutex{},
+		cache:           map[string][]dto.MetricFamily{},
+		lastUpdated:     time.Time{},
+	}, nil
+}
+
+func (f *KubernetesEndpointFetcher) FetchMetricsFor(ctx context.Context, endpoint string) ([]dto.MetricFamily, error) {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	if f.now().Sub(f.lastUpdated) < f.refreshInterval {
+		return f.cache[endpoint], nil
+	}
+
+	endpoints, err := f.discover(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	g, _ := errgroup.WithContext(ctx)
+	m := sync.Mutex{}
+	for _, ip := range endpoints {
+		ip := ip
+		g.Go(func() error {
+			metrics, err := fetchMetrics(f.client, f.buildAddr(ip), f.authToken)
+			if err != nil {
+				return err
+			}
+			m.Lock()
+			f.cache[ip] = metrics
+			m.Unlock()
+			return nil
+		})
```

Comment:
This might be premature optimization, but I did not want to find a bottleneck only once we run in production on a large cluster and need to query a lot of kubelets at the same time.

Reply:
Well AFAIK a goroutine will allocate 2KiB of stack memory.  If we assume a 1024 node cluster (which AFAIK we don't have)  that would be 2MiB of memory usage. There will probably be some overhead but we should still remain below 10MiB. I suspect the metrics cache will be larger. I think that's fine :sweat_smile: